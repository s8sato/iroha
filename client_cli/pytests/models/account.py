"""
This module provides an Account class for working with Iroha network accounts.
"""

from dataclasses import dataclass


@dataclass
class Account:
    """
    Account class represents an Iroha account.

    :param signatory: The signatory of the account.
    :type signatory: str
    :param domain: The domain of the account.
    :type domain: str
    """

    signatory: str
    domain: str

    def __repr__(self):
        return f"{self.signatory}@{self.domain}"
