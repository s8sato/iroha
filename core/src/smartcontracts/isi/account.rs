//! This module contains implementations of smart-contract traits and instructions for [`Account`] structure
//! and implementations of [`Query`]'s about [`Account`].

use iroha_data_model::{prelude::*, query::error::FindError};
use iroha_telemetry::metrics;

use super::prelude::*;
use crate::ValidQuery;

impl Registrable for iroha_data_model::account::NewAccount {
    type Target = Account;

    #[must_use]
    #[inline]
    fn build(self, _authority: &AccountId) -> Self::Target {
        self.into_account()
    }
}

/// All instructions related to accounts:
/// - minting/burning public key into account signatories
/// - minting/burning signature condition check
/// - update metadata
/// - grant permissions and roles
/// - Revoke permissions or roles
pub mod isi {
    use iroha_data_model::{
        asset::{AssetValue, AssetValueType},
        isi::{
            error::{MintabilityError, RepetitionError},
            InstructionType,
        },
        query::error::QueryExecutionFail,
    };
    use iroha_primitives::numeric::Numeric;

    use self::asset::isi::assert_numeric_spec;
    use super::*;
    use crate::{role::RoleIdWithOwner, state::StateTransaction};

    impl Execute for Register<Asset> {
        #[metrics(+"register_asset")]
        fn execute(
            self,
            authority: &AccountId,
            state_transaction: &mut StateTransaction<'_, '_>,
        ) -> Result<(), Error> {
            let asset_id = self.object.id;
            recognize_account(asset_id.account_id().clone(), authority, state_transaction)?;

            match state_transaction.world.asset(&asset_id) {
                Err(err) => match err {
                    QueryExecutionFail::Find(FindError::Asset(_)) => {
                        assert_can_register(
                            &asset_id.definition_id,
                            state_transaction,
                            &self.object.value,
                        )?;
                        let asset = state_transaction
                            .world
                            .asset_or_insert(asset_id.clone(), self.object.value)
                            .expect("account should exist");

                        match asset.value {
                            AssetValue::Numeric(increment) => {
                                state_transaction.world.increase_asset_total_amount(
                                    &asset_id.definition_id,
                                    increment,
                                )?;
                            }
                            AssetValue::Store(_) => {
                                state_transaction.world.increase_asset_total_amount(
                                    &asset_id.definition_id,
                                    Numeric::ONE,
                                )?;
                            }
                        }
                        Ok(())
                    }
                    _ => Err(err.into()),
                },
                Ok(_) => Err(RepetitionError {
                    instruction_type: InstructionType::Register,
                    id: IdBox::AssetId(asset_id.clone()),
                }
                .into()),
            }
        }
    }

    impl Execute for Unregister<Asset> {
        #[metrics(+"unregister_asset")]
        fn execute(
            self,
            _authority: &AccountId,
            state_transaction: &mut StateTransaction<'_, '_>,
        ) -> Result<(), Error> {
            let asset_id = self.object_id;

            let asset = state_transaction
                .world
                .account_mut(&asset_id.account_id)
                .and_then(|account| {
                    account
                        .remove_asset(&asset_id.definition_id)
                        .ok_or_else(|| FindError::Asset(asset_id))
                })?;

            match asset.value {
                AssetValue::Numeric(increment) => {
                    state_transaction
                        .world
                        .decrease_asset_total_amount(&asset.id.definition_id, increment)?;
                }
                AssetValue::Store(_) => {
                    state_transaction
                        .world
                        .decrease_asset_total_amount(&asset.id.definition_id, Numeric::ONE)?;
                }
            }

            state_transaction
                .world
                .emit_events(Some(AccountEvent::Asset(AssetEvent::Removed(
                    AssetChanged {
                        asset_id: asset.id,
                        amount: asset.value,
                    },
                ))));

            Ok(())
        }
    }

    impl Execute for Transfer<Account, AssetDefinitionId, Account> {
        fn execute(
            self,
            authority: &AccountId,
            state_transaction: &mut StateTransaction<'_, '_>,
        ) -> Result<(), Error> {
            let Transfer {
                source_id,
                object,
                destination_id,
            } = self;

            let _ = state_transaction.world.account(&source_id)?;
            recognize_account(destination_id.clone(), authority, state_transaction)?;

            let asset_definition = state_transaction.world.asset_definition_mut(&object)?;

            if asset_definition.owned_by != source_id {
                return Err(Error::Find(FindError::Account(source_id)));
            }

            asset_definition.owned_by = destination_id.clone();

            state_transaction
                .world
                .emit_events(Some(AssetDefinitionEvent::OwnerChanged(
                    AssetDefinitionOwnerChanged {
                        asset_definition_id: object,
                        new_owner: destination_id,
                    },
                )));

            Ok(())
        }
    }

    impl Execute for SetKeyValue<Account> {
        #[metrics(+"set_account_key_value")]
        fn execute(
            self,
            authority: &AccountId,
            state_transaction: &mut StateTransaction<'_, '_>,
        ) -> Result<(), Error> {
            let account_id = self.object_id;
            let account_metadata_limits = state_transaction.config.account_metadata_limits;
            recognize_account(account_id.clone(), authority, state_transaction)?;

            state_transaction
                .world
                .account_mut(&account_id)
                .expect("account should exist")
                .metadata
                .insert_with_limits(
                    self.key.clone(),
                    self.value.clone(),
                    account_metadata_limits,
                )
                .map_err(Error::from)?;

            state_transaction
                .world
                .emit_events(Some(AccountEvent::MetadataInserted(MetadataChanged {
                    target_id: account_id,
                    key: self.key,
                    value: self.value,
                })));

            Ok(())
        }
    }

    impl Execute for RemoveKeyValue<Account> {
        #[metrics(+"remove_account_key_value")]
        fn execute(
            self,
            _authority: &AccountId,
            state_transaction: &mut StateTransaction<'_, '_>,
        ) -> Result<(), Error> {
            let account_id = self.object_id;

            let value = state_transaction
                .world
                .account_mut(&account_id)
                .and_then(|account| {
                    account
                        .metadata
                        .remove(&self.key)
                        .ok_or_else(|| FindError::MetadataKey(self.key.clone()))
                })?;

            state_transaction
                .world
                .emit_events(Some(AccountEvent::MetadataRemoved(MetadataChanged {
                    target_id: account_id,
                    key: self.key,
                    value,
                })));

            Ok(())
        }
    }

    impl Execute for Grant<Permission, Account> {
        #[metrics(+"grant_account_permission")]
        fn execute(
            self,
            authority: &AccountId,
            state_transaction: &mut StateTransaction<'_, '_>,
        ) -> Result<(), Error> {
            let account_id = self.destination_id;
            let permission = self.object;
            let permission_id = permission.id.clone();

            recognize_account(account_id.clone(), authority, state_transaction)?;

            if !state_transaction
                .world
                .executor_data_model
                .permissions
                .contains(&permission_id)
            {
                return Err(FindError::Permission(permission_id).into());
            }

            if state_transaction
                .world
                .account_contains_inherent_permission(&account_id, &permission)
            {
                return Err(RepetitionError {
                    instruction_type: InstructionType::Grant,
                    id: permission.id.into(),
                }
                .into());
            }

            state_transaction
                .world
                .add_account_permission(&account_id, permission);

            state_transaction
                .world
                .emit_events(Some(AccountEvent::PermissionAdded(
                    AccountPermissionChanged {
                        account_id,
                        permission_id,
                    },
                )));

            Ok(())
        }
    }

    impl Execute for Revoke<Permission, Account> {
        #[metrics(+"revoke_account_permission")]
        fn execute(
            self,
            _authority: &AccountId,
            state_transaction: &mut StateTransaction<'_, '_>,
        ) -> Result<(), Error> {
            let account_id = self.destination_id;
            let permission = self.object;

            // Check if account exists
            state_transaction.world.account(&account_id)?;

            if !state_transaction
                .world
                .remove_account_permission(&account_id, &permission)
            {
                return Err(FindError::Permission(permission.id).into());
            }

            state_transaction
                .world
                .emit_events(Some(AccountEvent::PermissionRemoved(
                    AccountPermissionChanged {
                        account_id,
                        permission_id: permission.id,
                    },
                )));

            Ok(())
        }
    }

    impl Execute for Grant<RoleId, Account> {
        #[metrics(+"grant_account_role")]
        fn execute(
            self,
            authority: &AccountId,
            state_transaction: &mut StateTransaction<'_, '_>,
        ) -> Result<(), Error> {
            let account_id = self.destination_id;
            let role_id = self.object;

            let permissions = state_transaction
                .world
                .roles
                .get(&role_id)
                .ok_or_else(|| FindError::Role(role_id.clone()))?
                .clone()
                .permissions
                .into_iter()
                .map(|token| token.id);

            recognize_account(account_id.clone(), authority, state_transaction)?;

            if state_transaction
                .world
                .account_roles
                .insert(
                    RoleIdWithOwner::new(account_id.clone(), role_id.clone()),
                    (),
                )
                .is_some()
            {
                return Err(RepetitionError {
                    instruction_type: InstructionType::Grant,
                    id: IdBox::RoleId(role_id),
                }
                .into());
            }

            state_transaction.world.emit_events({
                let account_id_clone = account_id.clone();
                permissions
                    .zip(core::iter::repeat_with(move || account_id.clone()))
                    .map(|(permission_id, account_id)| AccountPermissionChanged {
                        account_id,
                        permission_id,
                    })
                    .map(AccountEvent::PermissionAdded)
                    .chain(std::iter::once(AccountEvent::RoleGranted(
                        AccountRoleChanged {
                            account_id: account_id_clone,
                            role_id,
                        },
                    )))
            });

            Ok(())
        }
    }

    impl Execute for Revoke<RoleId, Account> {
        #[metrics(+"revoke_account_role")]
        fn execute(
            self,
            _authority: &AccountId,
            state_transaction: &mut StateTransaction<'_, '_>,
        ) -> Result<(), Error> {
            let account_id = self.destination_id;
            let role_id = self.object;

            let permissions = state_transaction
                .world
                .roles
                .get(&role_id)
                .ok_or_else(|| FindError::Role(role_id.clone()))?
                .clone()
                .permissions
                .into_iter()
                .map(|token| token.id);

            if state_transaction
                .world
                .account_roles
                .remove(RoleIdWithOwner {
                    account_id: account_id.clone(),
                    role_id: role_id.clone(),
                })
                .is_none()
            {
                return Err(FindError::Role(role_id).into());
            }

            state_transaction.world.emit_events({
                let account_id_clone = account_id.clone();
                permissions
                    .zip(core::iter::repeat_with(move || account_id.clone()))
                    .map(|(permission_id, account_id)| AccountPermissionChanged {
                        account_id,
                        permission_id,
                    })
                    .map(AccountEvent::PermissionRemoved)
                    .chain(std::iter::once(AccountEvent::RoleRevoked(
                        AccountRoleChanged {
                            account_id: account_id_clone,
                            role_id,
                        },
                    )))
            });

            Ok(())
        }
    }

    /// Assert that this asset can be registered to an account.
    fn assert_can_register(
        definition_id: &AssetDefinitionId,
        state_transaction: &mut StateTransaction<'_, '_>,
        value: &AssetValue,
    ) -> Result<(), Error> {
        let expected_asset_value_type = match value.value_type() {
            AssetValueType::Numeric(_) => asset::isi::expected_asset_value_type_numeric,
            AssetValueType::Store => asset::isi::expected_asset_value_type_store,
        };
        let definition = asset::isi::assert_asset_type(
            definition_id,
            state_transaction,
            expected_asset_value_type,
        )?;
        if let AssetValue::Numeric(numeric) = value {
            assert_numeric_spec(numeric, &definition)?;
        }

        match definition.mintable {
            Mintable::Infinitely => Ok(()),
            Mintable::Not => Err(Error::Mintability(MintabilityError::MintUnmintable)),
            Mintable::Once => {
                if !value.is_zero_value() {
                    let asset_definition = state_transaction
                        .world
                        .asset_definition_mut(definition_id)?;
                    forbid_minting(asset_definition)?;
                    state_transaction.world.emit_events(Some(
                        AssetDefinitionEvent::MintabilityChanged(definition_id.clone()),
                    ));
                }
                Ok(())
            }
        }
    }

    /// Stop minting on the [`AssetDefinition`] globally.
    ///
    /// # Errors
    /// If the [`AssetDefinition`] is not `Mintable::Once`.
    #[inline]
    pub fn forbid_minting(definition: &mut AssetDefinition) -> Result<(), MintabilityError> {
        if definition.mintable == Mintable::Once {
            definition.mintable = Mintable::Not;
            Ok(())
        } else {
            Err(MintabilityError::ForbidMintOnMintable)
        }
    }

    #[cfg(test)]
    mod test {
        use iroha_data_model::{prelude::AssetDefinition, ParseError};
        use test_samples::gen_account_in;

        use crate::smartcontracts::isi::Registrable as _;

        #[test]
        fn cannot_forbid_minting_on_asset_mintable_infinitely() -> Result<(), ParseError> {
            let (authority, _authority_keypair) = gen_account_in("wonderland");
            let mut definition = AssetDefinition::numeric("test#hello".parse()?).build(&authority);
            assert!(super::forbid_minting(&mut definition).is_err());
            Ok(())
        }
    }
}

/// Account-related [`Query`] instructions.
pub mod query {

    use eyre::Result;
    use iroha_data_model::{
        account::Account, metadata::MetadataValueBox, permission::Permission,
        query::error::QueryExecutionFail as Error,
    };

    use super::*;
    use crate::state::StateReadOnly;

    impl ValidQuery for FindRolesByAccountId {
        #[metrics(+"find_roles_by_account_id")]
        fn execute<'state>(
            &self,
            state_ro: &'state impl StateReadOnly,
        ) -> Result<Box<dyn Iterator<Item = RoleId> + 'state>, Error> {
            let account_id = &self.id;
            state_ro.world().account(account_id)?;
            Ok(Box::new(
                state_ro.world().account_roles_iter(account_id).cloned(),
            ))
        }
    }

    impl ValidQuery for FindPermissionsByAccountId {
        #[metrics(+"find_permissions_by_account_id")]
        fn execute<'state>(
            &self,
            state_ro: &'state impl StateReadOnly,
        ) -> Result<Box<dyn Iterator<Item = Permission> + 'state>, Error> {
            let account_id = &self.id;
            Ok(Box::new(
                state_ro
                    .world()
                    .account_permissions_iter(account_id)?
                    .cloned(),
            ))
        }
    }

    impl ValidQuery for FindAllAccounts {
        #[metrics(+"find_all_accounts")]
        fn execute<'state>(
            &self,
            state_ro: &'state impl StateReadOnly,
        ) -> Result<Box<dyn Iterator<Item = Account> + 'state>, Error> {
            Ok(Box::new(
                state_ro
                    .world()
                    .domains_iter()
                    .flat_map(|domain| domain.accounts.values())
                    .cloned(),
            ))
        }
    }

    impl ValidQuery for FindAccountById {
        #[metrics(+"find_account_by_id")]
        fn execute(&self, state_ro: &impl StateReadOnly) -> Result<Account, Error> {
            let id = &self.id;
            iroha_logger::trace!(%id);
            state_ro
                .world()
                .map_account(id, Clone::clone)
                .map_err(Into::into)
        }
    }

    impl ValidQuery for FindAccountsByDomainId {
        #[metrics(+"find_accounts_by_domain_id")]
        fn execute<'state>(
            &self,
            state_ro: &'state impl StateReadOnly,
        ) -> Result<Box<dyn Iterator<Item = Account> + 'state>, Error> {
            let id = &self.domain_id;

            iroha_logger::trace!(%id);
            Ok(Box::new(
                state_ro.world().domain(id)?.accounts.values().cloned(),
            ))
        }
    }

    impl ValidQuery for FindAccountKeyValueByIdAndKey {
        #[metrics(+"find_account_key_value_by_id_and_key")]
        fn execute(&self, state_ro: &impl StateReadOnly) -> Result<MetadataValueBox, Error> {
            let id = &self.id;
            let key = &self.key;
            iroha_logger::trace!(%id, %key);
            state_ro
                .world()
                .map_account(id, |account| account.metadata.get(key).cloned())?
                .ok_or_else(|| FindError::MetadataKey(key.clone()).into())
                .map(Into::into)
        }
    }

    impl ValidQuery for FindAccountsWithAsset {
        #[metrics(+"find_accounts_with_asset")]
        fn execute<'state>(
            &self,
            state_ro: &'state impl StateReadOnly,
        ) -> Result<Box<dyn Iterator<Item = Account> + 'state>, Error> {
            let asset_definition_id = self.asset_definition_id.clone();
            iroha_logger::trace!(%asset_definition_id);

            Ok(Box::new(
                state_ro
                    .world()
                    .map_domain(&asset_definition_id.domain_id.clone(), move |domain| {
                        domain.accounts.values().filter(move |account| {
                            account.assets.contains_key(&asset_definition_id)
                        })
                    })?
                    .cloned(),
            ))
        }
    }
}
