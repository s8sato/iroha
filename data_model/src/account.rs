//! Structures, traits and impls related to `Account`s.
#[cfg(not(feature = "std"))]
use alloc::{collections::btree_map, format, string::String, vec::Vec};
use core::str::FromStr;
#[cfg(feature = "std")]
use std::collections::btree_map;

use derive_more::{Constructor, DebugCustom, Display};
use getset::Getters;
use iroha_data_model_derive::{model, IdEqOrdHash};
use iroha_schema::IntoSchema;
use parity_scale_codec::{Decode, Encode};
use serde::{Deserialize, Serialize};
use serde_with::{DeserializeFromStr, SerializeDisplay};

pub use self::model::*;
use crate::{
    asset::{
        prelude::{Asset, AssetId},
        AssetsMap,
    },
    domain::prelude::*,
    metadata::Metadata,
    HasMetadata, Identifiable, ParseError, PublicKey, Registered,
};

/// API to work with collections of [`Id`] : [`Account`] mappings.
pub type AccountsMap = btree_map::BTreeMap<AccountId, Account>;

#[model]
mod model {
    use super::*;

    /// Identification of [`Account`] by the combination of the [`PublicKey`] as its sole signatory and the [`Domain`](crate::domain::Domain) it belongs to.
    /// TODO #4373 include multi-signatory use.
    ///
    /// # Examples
    ///
    /// ```rust
    /// use iroha_data_model::account::AccountId;
    ///
    /// let id: AccountId = "ed0120BDF918243253B1E731FA096194C8928DA37C4D3226F97EEBD18CF5523D758D6C@domain".parse().expect("multihash@domain should be valid format");
    /// ```
    #[derive(
        DebugCustom,
        Display,
        Clone,
        PartialEq,
        Eq,
        PartialOrd,
        Ord,
        Hash,
        Constructor,
        Getters,
        Decode,
        Encode,
        DeserializeFromStr,
        SerializeDisplay,
        IntoSchema,
    )]
    #[display(fmt = "{signatory}@{domain_id}")]
    #[debug(fmt = "{signatory}@{domain_id}")]
    #[getset(get = "pub")]
    #[ffi_type]
    pub struct AccountId {
        /// [`Domain`](crate::domain::Domain) that the [`Account`] belongs to.
        pub domain_id: DomainId,
        /// Sole signatory of the [`Account`].
        pub signatory: PublicKey,
    }

    /// Account entity is an authority which is used to execute `Iroha Special Instructions`.
    #[derive(
        Debug,
        Display,
        Clone,
        IdEqOrdHash,
        Getters,
        Decode,
        Encode,
        Deserialize,
        Serialize,
        IntoSchema,
    )]
    #[allow(clippy::multiple_inherent_impl)]
    #[display(fmt = "({id})")] // TODO: Add more?
    #[ffi_type]
    pub struct Account {
        /// Identification of the [`Account`].
        pub id: AccountId,
        /// Assets in this [`Account`].
        pub assets: AssetsMap,
        /// Metadata of this account as a key-value store.
        pub metadata: Metadata,
    }

    /// Builder which should be submitted in a transaction to create a new [`Account`]
    #[derive(
        DebugCustom, Display, Clone, IdEqOrdHash, Decode, Encode, Serialize, Deserialize, IntoSchema,
    )]
    #[display(fmt = "[{id}]")]
    #[debug(fmt = "[{id:?}] {{ metadata: {metadata} }}")]
    #[ffi_type]
    pub struct NewAccount {
        /// Identification
        pub id: AccountId,
        /// Metadata that should be submitted with the builder
        pub metadata: Metadata,
    }
}

impl AccountId {
    /// Return `true` if the account signatory matches the given `public_key`.
    #[inline]
    #[cfg(feature = "transparent_api")]
    pub fn signatory_matches(&self, public_key: &PublicKey) -> bool {
        self.signatory() == public_key
    }
}

impl Account {
    /// Construct builder for [`Account`] identifiable by [`Id`] containing the given signatory.
    #[inline]
    #[must_use]
    pub fn new(id: AccountId) -> <Self as Registered>::With {
        <Self as Registered>::With::new(id)
    }

    /// Return a reference to the `Account` signatory.
    #[inline]
    pub fn signatory(&self) -> &PublicKey {
        &self.id.signatory
    }

    /// Return a reference to the [`Asset`] corresponding to the asset id.
    #[inline]
    pub fn asset(&self, asset_id: &AssetId) -> Option<&Asset> {
        self.assets.get(asset_id)
    }

    /// Get an iterator over [`Asset`]s of the `Account`
    #[inline]
    pub fn assets(&self) -> impl ExactSizeIterator<Item = &Asset> {
        self.assets.values()
    }
}

#[cfg(feature = "transparent_api")]
impl Account {
    /// Add [`Asset`] into the [`Account`] returning previous asset stored under the same id
    #[inline]
    pub fn add_asset(&mut self, asset: Asset) -> Option<Asset> {
        self.assets.insert(asset.id().clone(), asset)
    }

    /// Remove asset from the [`Account`] and return it
    #[inline]
    pub fn remove_asset(&mut self, asset_id: &AssetId) -> Option<Asset> {
        self.assets.remove(asset_id)
    }
}

impl NewAccount {
    fn new(id: AccountId) -> Self {
        Self {
            id,
            metadata: Metadata::default(),
        }
    }

    /// Add [`Metadata`] to the account replacing any previously defined metadata
    #[inline]
    #[must_use]
    pub fn with_metadata(mut self, metadata: Metadata) -> Self {
        self.metadata = metadata;
        self
    }
}

#[cfg(feature = "transparent_api")]
impl NewAccount {
    /// Convert into [`Account`].
    pub fn into_account(self) -> Account {
        Account {
            id: self.id,
            assets: AssetsMap::default(),
            metadata: self.metadata,
        }
    }
}

impl HasMetadata for NewAccount {
    fn metadata(&self) -> &Metadata {
        &self.metadata
    }
}

impl HasMetadata for Account {
    fn metadata(&self) -> &Metadata {
        &self.metadata
    }
}

impl Registered for Account {
    type With = NewAccount;
}

impl FromStr for AccountId {
    type Err = ParseError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.rsplit_once('@') {
            None => Err(ParseError {
                reason: "Account ID should have format `signatory@domain`",
            }),
            Some(("", _)) => Err(ParseError {
                reason: "Empty `signatory` part in `signatory@domain`",
            }),
            Some((_, "")) => Err(ParseError {
                reason: "Empty `domain` part in `signatory@domain`",
            }),
            Some((signatory_candidate, domain_id_candidate)) => {
                let signatory = signatory_candidate.parse().map_err(|_| ParseError {
                    reason: "Failed to parse `signatory` part in `signatory@domain`. `signatory` should have multihash format e.g. `ed0120...`",
                })?;
                let domain_id = domain_id_candidate.parse().map_err(|_| ParseError {
                    reason: "Failed to parse `domain` part in `signatory@domain`",
                })?;
                Ok(Self::new(domain_id, signatory))
            }
        }
    }
}

/// The prelude re-exports most commonly used traits, structs and macros from this crate.
pub mod prelude {
    pub use super::{Account, AccountId};
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_account_id() {
        const SIGNATORY: &str =
            "ed0120EDF6D7B52C7032D03AEC696F2068BD53101528F3C7B6081BFF05A1662D7FC245";
        let _ok = format!("{SIGNATORY}@domain")
            .parse::<AccountId>()
            .expect("should be valid");
        let _err_empty_signatory = "@domain"
            .parse::<AccountId>()
            .expect_err("@domain should not be valid");
        let _err_empty_domain = format!("{SIGNATORY}@")
            .parse::<AccountId>()
            .expect_err("signatory@ should not be valid");
        let _err_violates_format = format!("{SIGNATORY}#domain")
            .parse::<AccountId>()
            .expect_err("signatory#domain should not be valid");
    }
}
