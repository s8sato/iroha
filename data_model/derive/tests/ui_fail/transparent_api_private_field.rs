use iroha_data_model::account::AccountId;

fn main() {
    let account_id: AccountId = "ed0120CE7FA46C9DCE7EA4B125E2E36BDB63EA33073E7590AC92816AE1E861B7048B03@wonderland".parse().unwrap();
    println!("ID: {}", account_id.signatory);
}
