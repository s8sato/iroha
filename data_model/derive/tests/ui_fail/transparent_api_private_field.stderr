error[E0616]: field `signatory` of struct `iroha_data_model::account::AccountId` is private
 --> tests/ui_fail/transparent_api_private_field.rs:5:35
  |
5 |     println!("ID: {}", account_id.signatory);
  |                                   ^^^^^^^^^ private field
  |
help: a method `signatory` also exists, call it with parentheses
  |
5 |     println!("ID: {}", account_id.signatory());
  |                                            ++
